### ISSUE TITLE
[Bug] SM120 XQA MLA fp8: rowMaxLog2e metadata intermittently corrupted in the producer-to-consumer X handoff (masked today; blocks return_lse #3653)

### ISSUE THREAD

**Issue by waynehacking8** [flashinfer-ai/flashinfer#3839]
### Summary

On SM120 (RTX PRO 6000, CUDA 13), the fp8 XQA MLA decode kernel (`csrc/xqa/mla_sm120.cu`) intermittently delivers a corrupted `rowMaxLog2e` to the consumer CTAs, while the X payload and `rowSum` in the same `CgaXBuffer` transfer arrive intact. Nobody sees this today because nothing consumes the transferred row max in the shipping configuration: the single-block output normalization is invariant to the row max (numerator and denominator share the same `exp2(-M)`), and the only real consumers - the multi-block merge (`mergePartialOutputs`) and an LSE output - are respectively disabled (`launchMLAFlashInfer`: "MLA disables multi-block mode for now") and not implemented yet. It does mean the affected rows silently lose some precision, and it blocks both #3653 (`return_lse`) and any future multi-block re-enable, since the merge trusts the per-chunk `rowMaxLog2e`/`rowSum` pairs.

### How I found it

Implementing `return_lse` for #3653 (`lse = rowMaxLog2e + log2(rowSum)`, written from the consumer epilogue) made the corruption visible as huge finite LSE outliers. Against the fa2 MLA wrapper and a pure-torch logsumexp (which agree with each other to 4 decimals), the XQA LSE for single elements comes back at +26 to +213 in log2 units where the true value is ~9-10, on 1-4 elements out of ~100k per affected run, 1-4 affected cases per 396-case run of the existing fp8/bf16 test grids. Run-to-run on identical inputs: 10 of 39 repeat trials produced at least one differing LSE element (fa2 reference: 0 drift).

### Repro

Branch with the LSE detector: https://github.com/waynehacking8/flashinfer/tree/wayne/xqa-mla-return-lse (adds `return_lse` to `xqa_batch_decode_with_kv_cache_mla`).

Poison harness: fill the workspace scratch region with `0x7e` bytes before every call, so any read of not-yet-written scratch is loud. On the LSE branch (no other kernel changes), batch=64, page=128, seq lens randint(1,1024), fp8:

```python
ws = torch.zeros(128 * 1024 * 1024, dtype=torch.int8, device="cuda")
for t in range(600):
    ws[8 * 1024 * 1024:].fill_(0x7E)  # poison scratch (cgaXBuf area)
    _ = torch.randn(2048, 2048, device="cuda") @ torch.randn(2048, 2048, device="cuda")
    out, lse = flashinfer.decode.xqa_batch_decode_with_kv_cache_mla(
        query=query, kv_cache=kv_cache.unsqueeze(1), workspace_buffer=ws,
        qk_nope_head_dim=128, kv_lora_rank=512, qk_rope_head_dim=64,
        block_tables=block_tables, seq_lens=seq_lens_tensor, max_seq_len=max_seq_len,
        bmm1_scale=sm_scale, bmm2_scale=1.0, return_lse=True)
    assert lse.max().item() < 30  # true LSE is ~<=15 for these inputs
```

Measured rates (300-600 poisoned trials each, RTX PRO 6000, CUDA 13, main @ c53229ef + LSE patch):

| configuration | corruption events |
|---|---|
| fp8, LSE = rowMax + log2(rowSum) | 6/600, worst LSE 233.9 |
| fp8, LSE = rowMax only (debug) | 5/300, worst raw rowMax 165.7 |
| fp8, LSE = log2(rowSum) only (debug) | 0/300 |
| bf16, LSE = rowMax + log2(rowSum) | 0/300 |

So the corrupted datum is `rowMaxLog2e` specifically (the tail 512 B of the 9 KB `CgaXBuffer`), `rowSum` (the adjacent 512 B) stayed clean, and the bf16 kernel is clean under the identical harness. `enable_pdl` on/off are both affected. Everything is timing dependent: the same test passes in isolation and fails 1-4 cases when the full grid runs back-to-back.

### What I ruled out (each measured with the poison harness, 300-600 trials)

- `fence.proxy.async.shared::cta` before the `sendX` TMA store (the tma_store_fence pattern): no change.
- `fence.proxy.async.global` on both sides of the global `cgaXBuf` handoff: no change.
- `griddepcontrol.wait` at kernel entry: no change for this bug. (Separately: the kernel is launched with the PDL attribute but never calls `acqBulk()`, unlike `mha.cu`/`mha_sm90.cu` - that still looks worth fixing on its own.)
- 20 us delay between the consumer's produced-wait and its TMA load: no change, so it is not a simple store-completion latency race.

### Pointers

- The consumer-side handoff carries the original author's own doubt (`Consumer::loadX`): `// @fixme: check if this works. If it doesn't, randomly pick some data from dstX and dstRowSum and use STAS + arrive_tx to avoid fence.`
- The corrupted values look like plausible arithmetic results in the wrong scale regime (tens to low hundreds), not raw bit garbage, and in one poisoned run an event landed within 0.4 of `log2(poison_float)` - I could not reduce these observations to a single mechanism.

The LSE feature itself is ready on the branch above (fa2/torch-verified; the bf16 grid is 100% stable across repeated runs). I can send it either as a bf16-only PR now (fp8 + return_lse raising until this is fixed), or as the full PR once the metadata path is reliable - whichever you prefer. Happy to test candidate fixes on this card; it reproduces within a minute here.

### Environment

RTX PRO 6000 Blackwell Max-Q Workstation Edition (SM120), CUDA 13.0, driver 580.159.03, flashinfer main @ c53229ef, torch 2.11.0+cu130.

**Comment by waynehacking8** [comment]
Heads up for anyone landing here: the zip link in the comment above is malware spam (account created today, zero activity, link points at an unrelated throwaway repo) - don't download it. Hopefully a maintainer can remove it.

**Comment by Smallfu666** [comment]
A source-level hypothesis about the *class* of this bug, plus one candidate you can cross off. I have not reproduced it — no SM120 here — so everything below is read from `main` @ `668a1ba1` and nothing more. You have the only working repro, so what I can usefully offer is an argument that the bug is a different kind of thing than the four you eliminated, and a cheap experiment that decides that question regardless of which specific ordering flaw turns out to be responsible.

## The class of bug: wrong bytes, not late bytes

All four of your candidates (the two proxy fences, `griddepcontrol.wait`, the 20 µs consumer delay) change *when the bytes become visible*. None of them can help if the bytes placed in the buffer were already the wrong bytes. My suggestion is that this is a provenance problem: the transfer is working, and the consumer is receiving a `rowMaxLog2e` that belongs to a **different tile**.

Two of your observations point that way harder than they point at visibility:

**The values are plausible but in the wrong scale regime, not bit garbage** — your words. A running row max from a *later* tile is exactly that. And it is directionally right: the row max is accumulated monotonically along the ring (`csrc/xqa/mla_sm120.cu:921-923`, `lastRowMaxLog2e` loaded from shared memory then `quadRowMaxLog2e = fmaxf(tileRowMaxLog2e, replicateForQuad(..., lastRowMaxLog2e))`), so a value borrowed from later in the sequence is systematically *too large* — which is the direction all your outliers go. A visibility failure has no reason to prefer that direction.

**The 20 µs consumer-side delay changed nothing.** If the datum were merely late, that should have moved something. If the wrong value was already committed to global memory before the consumer looked, it explains itself.

Where the wrong bytes could come from, structurally:

- `:285-287` — there is **one** `CgaXBuffer x` in producer shared memory, time-shared by two math groups, with the comment `// single buffer reused by two groups. sendX() warp will arbitrate the order of x buffer access via two xBars.`
- `:884` — the global destination `args.cgaXBuf[nbSubSeq * idxInputTokenGlobal + idxSubSeq][ctaRank]` is the **same address every iteration** for a given (token, subSeq, ctaRank); the index has no `iter` term, so it is single-buffered too, guarded only by `cgaXBufConsumed`.
- `:211-215` — in `struct CgaXBuffer { XBuffer x; Vec<float,headGrpSize> rowSum; Vec<float,headGrpSize> rowMaxLog2e; }`, `rowMaxLog2e` is the **last** field. (Inference, not a measurement: `headGrpSize` is a compile-time macro, `mha.h:38`, and the MLA entry point rejects anything but 128 query heads — `flashinfer/xqa.py:620-624` — so 128 × 4 B = 512 B, which is what makes me read your "tail 512 B" as exactly this field. Worth a sanity check on your build.)
- `:815` — `rowMaxLog2e` is the **first** field the next tile's math warps write (`storeRowMax(smem.x.rowMaxLog2e, ...)`, ahead of `rowSum` at `:816` and the X payload at `:817`; same order in the bf16 branch at `:823-825`). Of the three fields in a reused buffer, it is the one written earliest and stored last.

So the structure would let a wrong-tile value reach the consumer if the buffer were ever writable a turn early. **But I could not find the flaw that would make that happen, and I would rather say so than assert a mechanism I cannot defend.** Two things I checked and could not turn into a bug:

- **The bulk store does drain before the buffer is released.** `tma::waitGroup<0>()` at `:887` emits plain `cp.async.bulk.wait_group 0` (`tma.h:252-254`), not the `.read` variant, so it waits on full completion of the committed group — strictly stronger than "the source shared memory has been read". The release to the other math group (`xBarNext.consumed.arrive()`, `:891`) is sequenced after it. So the naive "the store had not finished reading `smem.x` yet" race is **ruled out by the wait semantics** — worth stating plainly, since it was my first guess. (Elsewhere in the tree the weaker `.read` form is used deliberately where only source-read completion is needed — `csrc/fmha_v2/fmha/hopper/utils_tma.h:148` — so the distinction is clearly intentional.)
- **The ping-pong credit accounting looks correct to me.** The pre-arrive at `:871` grants group 0 its initial credit for an empty buffer, and thereafter `sendX` grants `xBars[(iter+1) % nbXBars].consumed` only after the store for `iter` has completed, so exactly one group should be writable at a time. I went looking for an extra or early release credit here and did not find one.

The two places I could not close statically, and where I would look next: the non-blocking fast path at `:792` (`bool const skipXBarWait = xBar.consumed.test_wait_parity(toParity<1>(grpIter));`, which bypasses the blocking `wait_parity` at `:812`, or `:820` on the bf16 side) — I see nothing wrong with it, but it is the one place where the WAR guard is satisfied by a poll rather than a wait; and the slot/parity mapping between the producer's `iter` (`:875`) and the consumer's `idxScratchXBuf = iter % nbProducerCtasPerCga` with `toParity<nbProducerCtasPerCga>(iter)` (`:1360`, `:1362`), where two producer CTAs feed one consumer's alternating slots.

So: reasonably confident about the *class* (wrong provenance), explicitly not confident about the specific flaw. Which is why the thing actually worth your time is the experiment below — it settles the class question without anyone having to guess the flaw first.

## The discriminator — mechanism-agnostic, and either outcome settles it

**Tag the payload with provenance.** Have the producer write `idxTile` into the transferred row max (or ship `rowMax + 1000*idxTile`), then on the consumer check whether a corrupted row decodes to a *different tile's* index.

- If it decodes to another tile → it is a provenance/lifetime bug, no fence will fix it, and the fix is structural: double-buffer `smem.x`, or give `cgaXBuf` per-iteration slots. That also tells you the search space is buffer ownership rather than memory ordering.
- If the index always matches the expected tile → the whole idea above is dead, and you have positively eliminated a class rather than tried one more fence.

A cheaper first probe if you want a five-minute signal before wiring up tagging: force `skipXBarWait = false` at `:792` so the blocking WAR wait is always taken, and re-run the harness. If the rate goes to 0/600 that fast path is implicated; if it does not budge, that is a clean strike against the buffer-reuse story.

## One candidate you can cross off

Worth stating explicitly so you do not spend a cycle on it: the missing-cluster-acquire theory does not hold, and I checked this one carefully because telling you to skip a live candidate would be the worst thing in this comment.

`cgaXBufProduced` is a `CgaBarrier` (`mla_sm120.cu:369`), and `barriers.cuh:398` defines `using CgaBarrier = MBarrier<Scope::CGA>`. The class is `template <Scope defaultScope_ = Scope::CTA> class MBarrier` with `static constexpr Scope defaultScope = defaultScope_;` at `barriers.cuh:57`, and `wait_parity`'s signature is `template <Scope scope = defaultScope>` (`:351`) — so for a `CgaBarrier` the scope resolves to `Scope::CGA`, and the consumer's wait at `mla_sm120.cu:1362` takes the `if constexpr (scope == Scope::CGA)` branch emitting `mbarrier.try_wait.parity.acquire.cluster` (`barriers.cuh:321`). That is correctly paired with the producer's explicit `fence.release.cluster` at `:892`, and both CTAs are in the same cluster, so cluster scope covers the global-window data. The release/acquire pair is already there and already correct.

## A possibility about `rowSum` that might be worth double-checking

This one I am genuinely unsure about, and it would be cheap to check while you are tagging things anyway. The "`rowSum` stayed clean" row comes from the `LSE = log2(rowSum)`-only build at 0/300. If the mechanism is buffer reuse rather than something specific to the row max, then `rowSum` would be getting swapped in from an adjacent tile *of the same run* — numerically plausible, and quite possibly inside the noise of a `lse.max() < 30` style check. Whereas a swapped running max shows up immediately, precisely because it grows monotonically across tiles. The same argument applies to the fp8-quantized X payload.

So it may be that the row max is the only field where corruption is *visible*, rather than the only field affected — which would matter, because "the corrupted datum is `rowMaxLog2e` specifically" naturally steers the search toward that one field and its fences. The same tagging experiment answers this too: tag `rowSum` as well and compare tile indices rather than magnitudes.

## Why bf16 being clean does not require a second mechanism

Worth noting, since it might otherwise look like evidence that this is fp8-specific: the bf16 `XBuffer` is twice the bytes (`mathElemBytes` 2 vs 1, `mla_sm120.cuh:45-47`), so a transfer of the same struct drains over a different window; and the bf16 branch skips the entire fp8 quantization block at `:795-811`, which sits between the softmax and the buffer wait. Any timing-sensitive race would be expected to move or close under both changes. Consistent with one mechanism rather than two.

## What I did not do

- **No repro on my side, at all.** No SM120 hardware here. Nothing above is measured — it is source reading against `668a1ba1` plus your report. Where I inferred rather than read something (the 512 B figure), I have said so.
- **No mechanism claim beyond the class.** As above: I could not identify the specific ordering flaw, and I ruled out my own first guess on the wait semantics.
- **One of your observations this does not explain**: the event that landed within 0.4 of `log2(poison_float)`. That looks like a genuine read of never-written scratch, which buffer reuse does not account for. The slot/parity mapping I mentioned (`:875` vs `:1360`/`:1362`) is where I would look, and it may be a second, rarer bug rather than the same one.
- I have not tried to write a fix and am not planning to — you have the repro and the detector branch.

Separately, your own observation that the kernel is launched with the PDL attribute but never calls `acqBulk()`, unlike `mha.cu` / `mha_sm90.cu`, looks like a real and probably cheap fix on its own merits, independent of this bug. That one seems worth splitting out whatever happens here.
